src/energy.h

```diff
@@ -242,575 +242,867 @@ class Bonded : public Energybase {
-                    auto v = words2vec<std::string>(i.key());
-                    if (v.size() == 2) {
-                        int id1 = (*findName(Faunus::molecules, v[0])).id();
-                        int id2 = (*findName(Faunus::molecules, v[1])).id();
-                        cutoff2.set(id1, id2, std::pow(i.value().get<double>(), 2));
+        }
+        return u;
+    }
+
+    /**
+     * @brief Complete cartesian pairing between a single particle in a group and particles in other groups in space.
+     *
+     * ⊕group × (space ∖ group), where ⊕ denotes a filter by an index (here a single particle)
+     *
+     * If the distance between the groups is greater or equal to the cutoff distance, the particle pairing between them
+     * is skipped. This method is performance-optimized version of the multiple indices method.
+     *
+     * @param group
+     * @param index a particle index relative to the group beginning
+     * @return energy sum between particle pairs
+     */
+    template <typename TGroup> double group2all(const TGroup &group, const int index) {
+        double u = 0;
+        const auto &particle = group[index];
+        for (auto &other_group : spc.groups) {
+            if (&other_group != &group) {                      // avoid self-interaction
+                if (!cut(other_group, group)) {                // check g2g cut-off
+                    for (auto &other_particle : other_group) { // loop over particles in other group
+                        u += particle2particle(particle, other_particle);
                     }
                 }
             }
         }
+        return u;
     }
 
     /**
-     * Calculates the force on all particles
-     * @todo Change to reflect only active particle, see Space::activeParticles()
+     * @brief Complete cartesian pairing between selected particles in a group and particles in other groups in space.
+     *
+     * ⊕group × (space ∖ group), where ⊕ denotes a filter by an index
+     *
+     * If the distance between the groups is greater or equal to the cutoff distance, the particle pairing between them
+     * is skipped.
+     *
+     * @param group
+     * @param index list of particle indices in the group relative to the group beginning
+     * @return energy sum between particle pairs
      */
-    void force(std::vector<Point> &forces) override {
-        auto &p = spc.p; // alias to particle vector (reference)
-        assert(forces.size() == p.size() && "the forces size must match the particle size");
-        for (size_t i = 0; i < p.size() - 1; i++) {
-            for (size_t j = i + 1; j < p.size(); j++) {
-                Point r = spc.geo.vdist(p[i].pos, p[j].pos); // minimum distance vector
-                Point f = pairpot.force(p[i], p[j], r.squaredNorm(), r);
-                forces[i] += f;
-                forces[j] -= f;
+    template <typename Tgroup> double group2all(const Tgroup &group, const std::vector<int> &index) {
```

Comment:
Actually the opposite is true. If `{index}` is used as an argument, the presumably faster `int index` overloaded method is called. After benchmarks will be created, this could be optimized using hard data.